### PR TITLE
drivers: clk: duty cycle, reparent pre-enable, linear rates description

### PR DESCRIPTION
**Pull request by etienne-lms**
Add an API function to get a clock duty cycle ratio.
Add an API function to get clock linear supported rates description.
Add a clock flag to pre-enable a parent clock before a clock a re-parented.
Update STM32MP13 clock driver.

### CODE REVIEW

**Comment by etienne-lms** [comment]
Rebased to get CI tests fixed.

**Comment by etienne-lms** [comment]
@GseoC, I've appended a commit for using on stm32mp13 clock get_duty_cycle operation that is introcued in this series.

**Comment by GseoC** [comment]
Hi @etienne-lms, I'll start the review after the build error is fixed in case you modify the implementation

**Comment by etienne-lms** [comment]
Oops, sorry! Done, I've fixed the appended commit.

**Comment by etienne-lms** [comment]
@GseoC, could you have a look at this series?

File: core/drivers/clk/clk.c [hunk 1]
```diff
@@ -205,9 +205,26 @@ static TEE_Result clk_set_rate_no_lock(struct clk *clk, unsigned long rate)
 	if (clk->parent)
 		parent_rate = clk_get_rate(clk->parent);
 
-	res = clk->ops->set_rate(clk, rate, parent_rate);
-	if (res)
-		return res;
+	assert(!(clk->flags & CLK_SET_RATE_PARENT) || clk->parent);
+	if (clk->flags & CLK_SET_RATE_PARENT) {
+		res = clk_set_rate_no_lock(clk->parent, rate);
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/clk/clk.c:210]: These is an assertion for that right above

File: core/include/drivers/clk.h [hunk 2]
```diff
@@ -199,6 +220,28 @@ TEE_Result clk_set_parent(struct clk *clk, struct clk *parent);
 TEE_Result clk_get_rates_array(struct clk *clk, size_t start_index,
 			       unsigned long *rates, size_t *nb_elts);
 
+/*
+ * clk_get_rates_steps - Get supported rates as min/max/step triplet
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/include/drivers/clk.h:224]: Ok, i'll address your concern.

**Comment by etienne-lms** [comment]
Review comment addressed.
I've rebased the series.

File: core/drivers/clk/clk.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (GseoC) [reply core/drivers/clk/clk.c:210]: Yes my bad, I got confused

**Comment by GseoC** [comment]
Ok to squash and apply `Reviewed-by: Gatien Chevallier <gatien.chevallier@foss.st.com>`

**Comment by jforissier** [comment]
> @jforissier: ok to merge, or should I rebase to latest master tip?

No need, that's fine.

**Pull request merged**